bridge: Support internal cgroupsv2 metrics

martinpitt opened the pull request:
We don't use this metric in cockpit itself for v2 - the docker page uses v1 (and its integration tests will make sure this does not regress). But I need it for https://github.com/martinpitt/performance-graphs to add the top CPU/memory consumers.

I added a smoke unit test, but the real functional test for now is something like this:

```
{ printf '\n{ "command": "open", "channel": "c", "payload": "metrics1", "source": "internal", "interval": 1000, "metrics": [ {"name": "cgroup.cpu.usage"} ] }\n---\n'; sleep 3; } | ./cockpit-bridge --interact=---
```

Fixes #14420

marusak commented:
Would this help with https://github.com/cockpit-project/cockpit-podman/issues/165 (or https://github.com/cockpit-project/cockpit-podman/pull/458) ? :crossed_fingers:

martinpitt commented:
@marusak: Yes, it would indeed. Of course you'd still need to add some heuristics to filter out which cgroups belong to podman containers, but that should be easy enough -- looks like filtering for "libpod-.*.scope" or "podman-.*.scope" shoud do the trick. (libpod-* looks like the pause container, and podman-* the real ones)

allisonkarlitskaya reviewed (approved):
To be honest, this looks really good.  My only suggestion is more of an opportunistic drive-by cleanup than an actual problem with this patch.

martinpitt commented on src/bridge/test-metrics.c:
```diff
@@ -719,6 +719,65 @@ test_deprecated_net_all (void)
+  while ((msg = mock_transport_pop_channel (transport, "1234")) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+  res = cockpit_json_parse_bytes (msg, NULL);
+  g_assert (res != NULL);
```
Added.

martinpitt commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -29,14 +29,14 @@
-read_double (const gchar *prefix,
-             const gchar *suffix)
+static gint64
+read_int64 (const gchar *prefix,
```
I had that in mind, then first shyed away from the whole rewrite. But it is so much more elegant, I changed the code now. Now the v1 code is much more consistent with the v2 one.

martinpitt commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -26,41 +26,65 @@
+  if (file_contents == NULL)
+      return -1;
+
+  offset = strstr (file_contents, value_prefix);
```
I went with a single read now. With cgroupsv2 the format is specified clearly (three fixed key names and long long values), so IMHO we can get away with a single 256 byte buffer and read.

martinpitt commented:
Updated for openat(), fixed a few places where values can be legitimately 0, and added the missing swap usage/max. I tested this on my cgroupsv2 fedora 32, and on v1 rhel 8.3.

martinpitt commented:
[this failure](https://logs.cockpit-project.org/logs/pull-14540-20200902-053946-c63d807b-fedora-33/log.html#246) is unrelated, but worth looking into. Looks like fedora-33 has unlearned memfd creation for restricted users?

allisonkarlitskaya commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,207 @@
+  const char *contents = read_file (dirfd, cgroup, fname);
+  if (contents == NULL)
+      return -1;
+  return atoll (contents);
```
We're talking to the kernel, of course, so I'm not sure how much error checking you want, ...

but maybe the file format changes or something?  We have warnings above for missing files...

martinpitt replied:
I added error checking for read_prefixed_int64(), but reverted it for general read_int64(). There are a lot of attributes whose value is "max" for example, and I don't want to hardcode a list of expected non-numbers, but just treat them as absent. So I went back to atoll() and ignoring a zero result where that is invalid.

allisonkarlitskaya commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,207 @@
-
-  path = g_build_filename (prefix, suffix, NULL);
-  if (!g_file_get_contents (path, &file_contents, &len, &error))
+  static char buf[256];
```
....you could also return a `FILE *` from this function :speak_no_evil:

martinpitt replied:
Did the first suggestion.

martinpitt commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,207 @@
+      g_warning ("cgroupfs value %s/%s is too large", cgroup, fname);
+      goto out;
+    }
+  buf[len] = '\0';
```
It's not, as `len == sizeof` is rejected. Not sure how to write that more clearly; would this comment help?
```c
  /* we really expect a much smaller read; if we get a full buffer, there's likely
   * more data, and we are misinterpreting stuff */
  if (len >= sizeof buf)
```

martinpitt commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,207 @@
-static double
-read_double (const gchar *prefix,
-             const gchar *suffix)
+static const char*
```
Fixed

allisonkarlitskaya commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,223 @@
-             const gchar *suffix)
+static const char *
+read_file (int dirfd,
+           char* buf,
```
hate to nag, but: `char *buf`

martinpitt replied:
Fixed.

allisonkarlitskaya commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,223 @@
+     user_usec 40000
+     system_usec 10000
+     */
+  val = read_prefixed_int64 (dirfd, cgroup, "cpu.stat", "usage_usec ");
```
Your approach here gives me pause.  What if some day they add a field called `max_usage_usec`?  Then this would match on that...

This is why I was leaning toward iterating over the lines...  I guess another thing you could do is to ensure that any match is either at the start of the buffer, or proceeded by a `\n` (and search for other matches, if not).

What do you think about this issue?

martinpitt replied:
Changed to checking the preceding char now, and renamed the function to match what it does.

allisonkarlitskaya commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,223 @@
-  prefix_len = strlen (prefix);
-
-  fs = fts_open ((gchar **)paths, FTS_NOCHDIR | FTS_COMFOLLOW, NULL);
+  fs = fts_open ((char **)paths, FTS_NOCHDIR | FTS_COMFOLLOW, NULL);
```
1. Cool.  I didn't know this class of functions existed!

2. Ugh.  A bit of an impedance mismatch here, now that we're in the `openat()` world.  That could be fixed with a followup, though.

martinpitt replied:
Would be nice if there was a walk function that spits out dirfds indeed, but I'm not aware of one.

allisonkarlitskaya commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,232 @@
+      if (offset == NULL)
+        return -1;
+      /* either matches at start of string, or after a line break */
+      if (offset == contents || offset[-1] == '\n')
```
This looks weird because `offset` is a bad name for this variable.  `offset` should be an int, but this is a pointer to a (potentially matching) key.  Might be worth a rename.

martinpitt replied:
Good point. Renamed to "match".

allisonkarlitskaya commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,232 @@
+      if (offset == contents || offset[-1] == '\n')
+        break;
+      offset += key_len;
+  }
```
So, I *think* this loop is going to do the right thing, but wow... I think we might be squarely in "making this complicated in order to optimise the wrong things" territory...

martinpitt replied:
Well, it was simpler before, but you didn't like it :-)

allisonkarlitskaya reviewed (approved):
I don't like the approach, but I won't stand in the way any longer :)

martinpitt commented on src/bridge/cockpitcgroupsamples.c:
```diff
@@ -21,119 +21,232 @@
+
+  /* search for a word match of key */
+  offset = contents;
+  for (;;) {
```
Argh, fixed.
